makePoplogCommander.sh

```diff
@@ -25,6 +27,92 @@ cat << \****
 #define PREFER_FLEXIBILITY  (RUN_INIT_P|INHERIT_ENV)
 
 
+///////////////////////////////////////////////////////////////////////////////
+//  Library routines: 
+//      Mishap, an error reporting function
+//      Chain, a 1D vector
+///////////////////////////////////////////////////////////////////////////////
+
+
+//  Mishap - Error reporting using printf-like functionality.
+
+void mishap( const char *msg, ... ) {
+    va_list args;
+    va_start( args, msg );
+    fprintf( stderr, "Mishap: " );
+    vfprintf( stderr, msg, args );
+    fprintf( stderr, "\n" );
+    va_end( args );
+    exit( EXIT_FAILURE );
+}
+
+//  Chains - 1D vectors.
+
+typedef struct Chain *Chain;
+typedef char * Ref;
+
+enum {
+    BUMP = 16
+};
+
+struct Chain {
+    int     size;
+    int     used;
+    Ref     *data;
+};
+
+// Ensure there is room for at least n more bytes
+// in the chain's buffer.
+//
+static Chain bump( Chain r, int n ) {
+    int size = r->size;
+    int used = r->used;
+    int newused = used + n;
+    if ( newused > size ) {
+        int newsize = newused + BUMP;
```

Comment:
What is the purpose of `BUMP`? It seems you're allocating an extra 16 elements worth of free space in the chain over the requested `n`. If so, why 16, what's the logic for it, can this be added as a comment on the definition of `BUMP`?

Reply:
Short answer: it is a bad idea to only request the amount of store needed, as reallocation is at least O(N). In the common case where you push single items onto a vector, this leads to O(N^2) performance. The proper answer to tackle this is actually to grow the length exponentially and the usual value is a factor of 2. However that factor of 2 has terrible performance at small sizes and introduces a new corner case that the 0-size vector cannot be grown. Also a factor of 2 grows the vector too fast in practice and leads to an even higher level of store waste than normal (that's a big topic but C's "efficient" store management is eventually wasteful because of fragmentation plus extra waste because it usually uses a buddy system). In this very limited case, it's good enough to grow the size by a linear amount because the number of `--option` parameters is unlikely to be in the 1000s.

Having written all of this, I will fix the growth algorithm!

Reply:
Fixed.